pkg/provider/azure_wrap.go

```diff
@@ -125,6 +126,52 @@ func (az *Cloud) getPublicIPAddress(pipResourceGroup string, pipName string, crt
 	return *(deepcopy.Copy(pip).(*network.PublicIPAddress)), true, nil
 }
 
+// getPublicIPAddressWithRetry gets public IP address with provided pipName and its non-IPFamily version.
+func (az *Cloud) getPublicIPAddressWithRetry(pipResourceGroup string, pipNameWithIPFamily string, crt azcache.AzureCacheReadType) (network.PublicIPAddress, bool, error) {
+	pipNameNoIPFamily, isIPv6 := getResourceNoIPFamily(pipNameWithIPFamily)
+
+	cached, err := az.pipCache.Get(pipResourceGroup, crt)
+	if err != nil {
+		return network.PublicIPAddress{}, false, err
+	}
+
+	pips := cached.(*sync.Map)
+	var pip any
+	var okWithIPFamily, okNoIPFamily bool
+	pip, okWithIPFamily = pips.Load(pipNameWithIPFamily)
+	if !okWithIPFamily {
```

Comment:
A corner case: suppose a service has xxx and xxx-IPv6 existing but only xxx in cache. When running this function for the ipv6 version, xxx-IPv6 is not found in cache and then xxx is found. But since version does not match, this function returns notFound at line 169. Looks like we need to do the cache refresh when pipNameWithIPFamily is not found but not when pipNameNoIPFamily is not found.

Reply:
Could you please share how to achieve such situation with xxx (IPv4) PIP in cache but xxx (IPv4) and xxx-IPv6 PIPs exist?
If cache refresh earlier, ccm will always do cache refresh when a user upgrades ccm to a version with dualstack supported. API call will increase a lot.

Reply:
The situation I can come up with is: Since xxx exists, it is an old cluster and it upgrades ccm to a version supporting dualstack. Then the user turns the cluster from a IPv4 one into a dualstack one and creates a dualstack Service so xxx-IPv6 is created. However, `CreateOrUpdatePIP()` has `az.pipCache.Delete(pipResourceGroup)`. As a result, I don't think xxx in cache but xxx-IPv6 not will happen? Please correct me if I'm wrong.

Reply:
how about changing the service IPs from single stack to dual stack? e.g. ipv4<->dualstack and ipv6<->dualstack.